docs/reference-guide/modules/events/pages/event-versioning.adoc

```diff
@@ -3,45 +3,166 @@
 In the lifecycle of an Axon application, their events _will_ typically change their format.
 As events are stored indefinitely, the application should be able to cope with several versions of an event.
 This chapter will discuss what to keep in mind when creating your events, for backwards (and forwards) compatibility.
-It will also explain the upcasting process.
+It will explain Axon's versioning approach using payload conversion and when you might need upcasters.
+
+== Event versioning approach
+
+Axon Framework provides a flexible approach to event versioning that minimizes the need for complex migration strategies.
+The primary mechanism is **payload conversion at handling time**, which allows different handlers to receive events in the format they expect.
+
+=== Payload conversion at handling time
+
+When events are stored in the event store, they retain their serialized form (for example JSON or Avro binary) along with metadata about their `MessageType`.
+When a handler needs to process an event, Axon converts the payload to the handler's expected type at that moment.
+
+This means:
+
+* Different handlers can process the same event using different Java representations
+* You can evolve your event handling without modifying stored events
+* Simple structural changes often require no upcasters
+
+For example, consider an event stored with this structure:
+
+[source,java]
+----
+@Event(name = "OrderPlaced", version = "1.0.0")
+public class OrderPlacedEvent {
+    private final String orderId;
+    private final String customerId;
+    private final List<String> productIds;
+
+    // constructors, getters...
+}
+----
+
+Later, you want to process this same event with additional computed information:
+
+[source,java]
+----
+public class EnrichedOrderPlacedEvent {
+    private final String orderId;
+    private final String customerId;
+    private final List<String> productIds;
+    private final int productCount;  // Computed from productIds
+
+    // Constructor that computes productCount
+    public EnrichedOrderPlacedEvent(String orderId, String customerId, List<String> productIds) {
+        this.orderId = orderId;
+        this.customerId = customerId;
+        this.productIds = productIds;
+        this.productCount = productIds != null ? productIds.size() : 0;
+    }
+
+    // getters...
+}
+----
+
+You can add a handler that receives the event in this new format:
+
+[source,java]
+----
+@EventHandler(eventName = "com.example.events.OrderPlaced")
```

Comment:
For this example to make sense, the `eventName` of the old event handler and the new need to be the same. They're not right now, so we should change that.